virtualhostkeys.go

```
@@ -0,0 +1,125 @@
+	if err != nil {
+		return nil, errors.Annotatef(err, "virtual host keys version schema check failed")
+	}
+	valid := coerced.(map[string]interface{})
+
+	version := int(valid["version"].(int64))
+	sourceList := valid["virtual-host-keys"].([]interface{})
+	return importVirtualHostKeyList(sourceList, version)
+}
+
+func importVirtualHostKeyList(sourceList []interface{}, version int) ([]*virtualHostKey, error) {
+	getFields, ok := virtualHostKeyFieldsFuncs[version]
+	if !ok {
+		return nil, errors.NotValidf("version %d", version)
+	}
+
+	result := make([]*virtualHostKey, 0, len(sourceList))
+	for i, value := range sourceList {
+		source, ok := value.(map[string]interface{})
+		if !ok {
+			return nil, errors.Errorf("unexpected value for virtual host keys %d, %T", i, value)
+		}
+		virtualHostKey, err := importVirtualHostKey(source, version, getFields)
+		if err != nil {
+			return nil, errors.Annotatef(err, "virtual host keys %d", i)
+		}
+		result = append(result, virtualHostKey)
+	}
+	return result, nil
+}
+
+var virtualHostKeyFieldsFuncs = map[int]fieldsFunc{
+	1: virtualHostKeyV1Fields,
+}
+
+func virtualHostKeyV1Fields() (schema.Fields, schema.Defaults) {
+	fields := schema.Fields{
+		"id":       schema.String(),
+		"host-key": schema.String(),
+	}
+	return fields, schema.Defaults{}
+}
+
+func importVirtualHostKey(source map[string]interface{}, importVersion int, fieldFunc func() (schema.Fields, schema.Defaults)) (*virtualHostKey, error) {
+	fields, defaults := fieldFunc()
+	checker := schema.FieldMap(fields, defaults)
+
+	coerced, err := checker.Coerce(source, nil)
+	if err != nil {
+		return nil, errors.Annotatef(err, "virtual host keys v%d schema check failed", importVersion)
+	}
+	valid := coerced.(map[string]interface{})
+	// From here we know that the map returned from the schema coercion
+	// contains fields of the right type.
+
+	consumer := &virtualHostKey{
+		ID_:      valid["id"].(string),
+		HostKey_: valid["host-key"].(string),
```

If the decode error is being ignored, We should check here to be safe.
I would argue we could consider
```
type virtualHostKey struct {
	ID_      string  `yaml:"id"`
	HostKeyEncoded_ string `yaml:"host-key"`
        HostKey_ string `yaml:"-"`
}
```
The HostKey() getter returns `HostKey_` which we always know to be valid. The encoded value is sent over the wire and always decoded during import where the error can be handled.

---

I went and checked the other usage of base64 encoding in the repo and that's in application.MetricsCredentials.
I see there it checks the value can be decoded during imported. So I've done the same here.
```
	// The model stores the virtual host key encoded, but we want to
	// make sure that we are storing something that can be decoded.
	if _, err := base64.StdEncoding.DecodeString(hostKey); err != nil {
		return nil, errors.Annotate(err, "virtual host key not valid")
	}
```

---

Yup, that matches the other thought I had about it. Good to be consistent.